streams/src/test/java/org/apache/kafka/streams/KafkaStreamsTest.java

```diff
@@ -982,10 +983,14 @@ public void shouldThrowOnCleanupWhileShuttingDownStreamClosedWithCloseOptionLeav
             closeOptions.timeout(Duration.ZERO);
             closeOptions.leaveGroup(true);
 
-            streams.close(closeOptions);
-            assertThat(streams.state() == State.PENDING_SHUTDOWN, equalTo(true));
-            assertThrows(IllegalStateException.class, streams::cleanUp);
-            assertThat(streams.state() == State.PENDING_SHUTDOWN, equalTo(true));
+            if (streams.close(closeOptions)) {
```

Comment:
I don't think this fixes the problem. After `streams.close()` returned, and did return `false`, there might be another context switch transiting the state to NOT_RUNNING.

We really need to _block_ the `CleanupThread` to make progress. We actually try to do this via `prepareTerminableThread` and let the thread sleep for 2 seconds to block. Seems that is not strong enough.

I might be better to actually use an `AtomicBoolean` that would block the `join()` call until we flip it at the end of the test.

Reply:
Yes, I think you are right. We should drop this PR and find a solution in the `CleanupThread` side

Reply:
Not sure if I agree -- `CleanupThread` seem to be fine, but it seem to rather be a test setup issue, ie, we need a clean way to "pause" the `CleanupThread` for test -- for regular production code, we don't need to pause it though, so it's only a test setup thing.

As mentioned, I think if we change `prepareTerminableThread` helper method of this test, we can make it work. If it's unclear to you what I have in mind, happy to do a PR myself. Just let me know how you want to proceeed.

Reply:
Ah. Just saw https://github.com/apache/kafka/pull/18418 :)